From 1886c082c93d27dc6e8ae73a2696b01f3a5e1d74 Mon Sep 17 00:00:00 2001
From: zhang sanshan <b51434@freescale.com>
Date: Thu, 5 Nov 2015 13:01:10 +0800
Subject: [PATCH 18/21] MA-7218-2 [Android_imx6_auto]factoryreset:do factory
 reset with binay ubiupdatevol in recovery

This patch call binary ubiupdatevol to do factory reset.

Signed-off-by: zhang sanshan <b51434@freescale.com>
---
 Android.mk         |  1 +
 ubi.cpp            | 64 +++++++++++++++++++++++++++---------------------------
 updater/Android.mk |  2 +-
 3 files changed, 34 insertions(+), 33 deletions(-)

diff --git a/Android.mk b/Android.mk
index 879ef48..a1d542e 100644
--- a/Android.mk
+++ b/Android.mk
@@ -92,6 +92,7 @@ LOCAL_C_INCLUDES += external/mtd-utils/ubi-utils/include external/mtd-utils/incl
 LOCAL_STATIC_LIBRARIES += libubi
 LOCAL_SRC_FILES += ubi.cpp
 
+LOCAL_REQUIRED_MODULES := ubiupdatevol
 ifeq ($(HAVE_SELINUX), true)
   LOCAL_C_INCLUDES += external/libselinux/include
   LOCAL_STATIC_LIBRARIES += libselinux
diff --git a/ubi.cpp b/ubi.cpp
index 6ae72fc..9595184 100644
--- a/ubi.cpp
+++ b/ubi.cpp
@@ -1,5 +1,5 @@
 /*
- * Copyright (C) 2014 Freescale Semiconductor, Inc.
+ * Copyright (C) 2014-2015 Freescale Semiconductor, Inc.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -18,42 +18,42 @@
 #include <fcntl.h>
 #include <stdio.h>
 #include <sys/stat.h>
+#include <sys/types.h>
+#include <sys/wait.h>
 #include <libubi.h>
+#include <cutils/log.h>
+#include<unistd.h>
+
+
+static int exec_cmd(const char* path, char* const argv[]) {
+    int status;
+    pid_t child;
+    int ret;
+    if ((child = vfork()) == 0) {
+        ret = execv(path, argv);
+	if (ret < 0)
+		printf("exec_cmd errno is %d\n",errno);
+        _exit(0);
+    }
+    waitpid(child, &status, 0);
+    if (!WIFEXITED(status) || WEXITSTATUS(status) != 0) {
+        printf("%s failed with status %d\n", path, WEXITSTATUS(status));
+    }
+    return WEXITSTATUS(status);
+}
 
 extern "C" int ubiVolumeFormat(char *location)
 {
     int ret = -1;
-    libubi_t libubi;
-    struct ubi_vol_info vol_info;
-    int fd;
-    libubi = libubi_open();
-
-    if (libubi == NULL) {
-        fprintf(stderr, "can not open libubi");
-        goto done;
-    }
-
-    ret = ubi_get_vol_info(libubi, location, &vol_info);
-    if (ret) {
-        fprintf(stderr, "can not get information about UBI volume %s", location);
-        goto done;
-    }
-
-    fd = open(location, O_RDWR);
-    if (fd == -1) {
-        fprintf(stderr, "can not open %s", location);
-        goto done;
-    }
-
-    ret = ubi_update_start(libubi, fd, 0);
-    if (ret) {
-        fprintf(stderr, "cannot truncate volume %s", location);
-        close(fd);
-        goto done;
-    }
-
-    close(fd);
-
+    const char *ubiupdatevol = "/sbin/ubiupdatevol";
+    const char* args[] = {"ubiupdatevol", location, "-t", NULL};
+    printf("Formatting %s\n",location);
+
+    ret = exec_cmd(ubiupdatevol, (char* const*)args);
+    if (ret != 0) {
+	printf("format_volume %s failed\n",location);
+	return -1;
+    }	
   done:
     return ret;
 }
diff --git a/updater/Android.mk b/updater/Android.mk
index ae9786e..e389ed9 100644
--- a/updater/Android.mk
+++ b/updater/Android.mk
@@ -84,7 +84,7 @@ LOCAL_CFLAGS += -DUSE_UBIFS
 LOCAL_C_INCLUDES += external/mtd-utils/ubi-utils/include/ ./external/mtd-utils/include/
 LOCAL_STATIC_LIBRARIES += libubi
 LOCAL_SRC_FILES += ../ubi.cpp
-
+LOCAL_REQUIRED_MODULES := ubiupdatevol
 inc :=
 inc_dep_file :=
 
-- 
1.8.0

